### ISSUE TITLE
Analyzer to guard against comparing struct to null

### ISSUE THREAD

**Issue by AmadeusW** [dotnet/roslyn-analyzers#3986]
### Describe the problem you are trying to solve

We checked if a variable of struct type is null. The check was negative, and we took incorrect code path.
We should have checked if the variable is default. This would have resulted in positive check, and we would have taken correct code path.

I would like to guard against such with an analyzer.
Does an analyzer like this already exist?

### Describe suggestions on how to achieve the rule

1. Look for comparison between value type and null (`== null` or `!= null`)
2. The fix is to replace `null` with `default`

I guess implementation could be similar to one of [CA2242](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2242?view=vs-2019)

### Additional context

Sample scenario
```csharp

struct MyStructWithObject
{
    object MyObject {get;}
    public myStruct(object o)
    {
        MyObject = o ?? throw new ArgumentNullException(nameof(o));
    }
}

// Code which triggers NullReferenceException
GetText(default);

// Analyzer should raise a warning in this method
string GetText(MyStructWithObject s)
{
    if (s == null) // Evaluates to false, branch not taken
    {
        return string.Empty;
    }
    return s.MyObject.ToString(); // Throws NullReferenceException
}

// Fixed
string GetText(MyStructWithObject s)
{
    if (s == default) // Fixed
    {
        return string.Empty;
    }
    return s.MyObject.ToString();
}
```

**Comment by mavasani** [comment]
Tagging @jmarolf @gafter @jaredpar - I believe this is one of the diagnostic planned for compiler warning wave. If not, this should certainly be a good candidate for warning level analyzer diagnostic that ships with the .NET SDK.

**Comment by jmarolf** [comment]
I believe this is exactly encompasses https://github.com/dotnet/roslyn/issues/45744 and is already checked into roslyn master

**Comment by mavasani** [comment]
Great, closing this as external!

**Comment by AmadeusW** [comment]
Thanks @jmarolf! I read through the PR you linked and its linked PRs, and I'm curious how to enable this.

* Will compiler shipping with VS 16.8 allow me to use `<AnalysisLevel>5</AnalysisLevel>`?
* Is there a way to opt into seeing CS8073 on a build machine which hasn't been updated to VS 16.8? Is `<Features>strict</Features>`the way to go?
* Additionally, do I need to enable nullability to see CS8073?

**Comment by jmarolf** [comment]
`<WarningLevel>5</WarningLevel>` is probably the way to go. I have no idea how CoreXT targets work so I am unable to speculate what other msbuild options will function correctly. `<AnalysisLevel>5</AnalysisLevel>` is an alias for this that also enable .NET platform analyzers. If all you care about is getting the new warnings from the compiler setting `WarningLevel` to 5 should do what you want.
